src/inline-assembly.md

```diff
@@ -37,515 +50,892 @@ assert_eq!(x, 4 * 6);
+A `format_specifier` that has an `operand_specifier` which is a DEC_LITERAL is called an explicit positional specifier, and refers to the `nth` successive positional operand, where `n` is the value of the DEC_LITERAL.
+
+```rust
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("mov {1}, {0}", in(reg) 5i64, out(reg) x);
+# }}
+```
+
+r[asm.invocation.named]
+A `format_specifier` that has an `operand_specifier` which is an ident is called a named specifier, and refers to the named operand with the specified name.
+
+```rust
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("mov {output}, {input}", input = in(reg) 5i64, output = out(reg) x);
+# }}
+```
+
+r[asm.invocation.expansion]
+The *joined asm-string* is expanded as defined in [asm.operands.expansion], replacing each `format_specifier` with the appropriate expansion for the operand. The resulting string is called the *expanded asm-string*.
+
+r[asm.invocation.syntax]
+The syntax of the *expanded asm-string* is a subset of the GNU AS syntax for the target. Invoking the macro with a *expanded asm-string* that does not match syntax requirements is *conditionally supported* and has *assembler dependent behaviour*. Invoking a directive that is not specified by [asm.directives] is *conditionally supported* and has *assembler dependent behaviour*.
+
+> [!TARGET-SPECIFIC]
+> On x86 and x86_64 targets, the syntax of the *expanded asm-string* acts as though the directive `.intel_syntax noprefix` is issued before parsing the *expanded asm-string*, except that the `option(att_syntax)` causes the syntax to act as though the directive `.att_syntax prefix` is issued before parsing the *expanded asm-string* instead.
+> On ARM and Aarch64 targets, the syntax of the *expanded asm-string* acts as though the directive `.syntax unified` is issued before parsing the *expanded asm-string*.
+
+r[asm.invocation.duplication]
+The number of times, locations, and the order in which a given invocation of [`core::arch::asm!`] is expanded is unspecified.
+
+```rust,ignore
+// The following code may have suprising results, and may fail to compile or link. 
+// The results, including whether it succesfully compiles, may depend on non-local use sites of the function, and on optimization settings.
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("foo: jmp foo");
+# }}
+```
+
+> [!NOTE]
+> In particular, an asm block may be duplicated, for example if the containing function is inlined, or omitted from the output entirely.
+> As a consequence, asm blocks should not use directives that have non-idempotent non-local effects, or named labels and symbol definitions. 
+> Additionally, two asm blocks may not rely upon being adjacent in executable memory, even if they are adjacent in the source.
+
+> [!NOTE]
+> Local Labels (a decimal literal that doesn't solely consist of 0s and 1s) may be used freely if the asm block needs to define a label.
```

Comment:
I think it should be noted so that people don't rely on the statement to say "But it says 'a decimal literal'", but it could be mentioned as being a bug. We can always loosen the specification, but we can't tighten it later after everything is all stabilized.